docs/smart-contracts/meebits_erc721.sol

```diff
@@ -0,0 +1,679 @@
+        }
+        bytes memory buffer = new bytes(digits);
+        uint256 index = digits - 1;
+        temp = value;
+        while (temp != 0) {
+            buffer[index--] = bytes1(uint8(48 + temp % 10));
+            temp /= 10;
+        }
+        return string(buffer);
+    }
+
+    /**
+      * @dev Returns a descriptive name for a collection of NFTokens.
+      * @return _name Representing name.
+      */
+    function name() external view returns (string memory _name) {
+        _name = nftName;
+    }
+
+    /**
+     * @dev Returns an abbreviated name for NFTokens.
+     * @return _symbol Representing symbol.
+     */
+    function symbol() external view returns (string memory _symbol) {
+        _symbol = nftSymbol;
+    }
+
+    /**
+     * @dev A distinct URI (RFC 3986) for a given NFT.
+     * @param _tokenId Id for which we want uri.
+     * @return _tokenId URI of _tokenId.
+     */
+    function tokenURI(uint256 _tokenId) external view validNFToken(_tokenId) returns (string memory) {
+        return string(abi.encodePacked("https://meebits.larvalabs.com/meebit/", toString(_tokenId)));
+    }
+
+    //// MARKET
+
+    struct Offer {
+        address maker;
+        address taker;
+        uint256 makerWei;
+        uint256[] makerIds;
+        uint256 takerWei;
+        uint256[] takerIds;
+        uint256 expiry;
+        uint256 salt;
+    }
+
+    function hashOffer(Offer memory offer) private pure returns (bytes32){
+        return keccak256(abi.encode(
+                    offer.maker,
+                    offer.taker,
+                    offer.makerWei,
+                    keccak256(abi.encodePacked(offer.makerIds)),
+                    offer.takerWei,
+                    keccak256(abi.encodePacked(offer.takerIds)),
+                    offer.expiry,
+                    offer.salt
+                ));
+    }
+
+    function hashToSign(address maker, address taker, uint256 makerWei, uint256[] memory makerIds, uint256 takerWei, uint256[] memory takerIds, uint256 expiry, uint256 salt) public pure returns (bytes32) {
+        Offer memory offer = Offer(maker, taker, makerWei, makerIds, takerWei, takerIds, expiry, salt);
+        return hashOffer(offer);
+    }
+
+    function hashToVerify(Offer memory offer) private pure returns (bytes32) {
+        return keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", hashOffer(offer)));
+    }
+
+    function verify(address signer, bytes32 hash, bytes memory signature) internal pure returns (bool) {
+        require(signer != address(0));
+        require(signature.length == 65);
+
+        bytes32 r;
+        bytes32 s;
+        uint8 v;
+
+        assembly {
+            r := mload(add(signature, 32))
+            s := mload(add(signature, 64))
+            v := byte(0, mload(add(signature, 96)))
+        }
+
+        if (v < 27) {
+            v += 27;
+        }
+
+        require(v == 27 || v == 28);
+
+        return signer == ecrecover(hash, v, r, s);
+    }
```

Comment:
Assume this is used to validate that the signature (on chain) matches what is in the offer